core/domain/user_services.py

```diff
@@ -111,7 +119,27 @@ def require_valid_username(cls, username):
             # Admin usernames are reserved for admins. Note that 'admin'
             # itself is already in use for the demo exploration.
             raise utils.ValidationError('This username is not available.')
-
+    @classmethod
+    def fetch_gravatar(cls, email):
+        """Gets autogenerated avatar for a given email.
+
+        Returns Default picture if there is an error.
+        """
+        base_url = "http://www.gravatar.com/avatar/"
+        avatar_url = base_url + hashlib.md5(email.lower()).hexdigest() + "?"
+        avatar_url += urllib.urlencode({'d':'retro', 's':str(AVATAR_SIZE)})
```

Comment:
Interesting, I didn't realize gravatar provided identicons. Thanks for the screenshot!

Given this, I think that, without the benchmarking to show that generating images on the fly is faster, I would lean towards grabbing the identicon from gravatar and storing it, if the user doesn't already have a 'proper' gravatar picture. If any HTTP (or other) errors occur, we can still store None.